### ISSUE TITLE
Dont trim white space by default

### ISSUE THREAD

**Issue by alphaCTzo7G** [python-mode/python-mode#912]
It seems `python-mode` turns on trimming white space by default. I do like this option when I am working on my own repository. 

However, if you work with distributed repositories, this will frequently delete white spaces, which will cause a lot more changes in the pull-requests.

Its probably better to set

`let g:pymode_trim_whitespace=0` by default. 

What do you guys think? If its ok.. I can do a PR.

**Comment by alphaCTzo7G** [comment]
Heres' the line: https://github.com/python-mode/python-mode/blob/bb746d0d0cba9adedbac856429e37a0dbfc599c6/plugin/pymode.vim#L55

**Comment by diraol** [comment]
@alphaCTzo7G I do also prefer my settings with this feature enabled. I do understand the problem while contributing or working with code that is not that clear.

I am not sure if there is any recommendation on PEP's about this.

If there is, I am for keeping it as it is today because of that.

If not, this seems to be a best practice, so I think it is a good idea for us to keep it like that in order to educate newcomers and other developers about keeping their code clean.

In any case, one can always disable this option.

What do you think?

**Comment by alphaCTzo7G** [comment]
I have to look up what PEP says.. but I actually came across this issue while contributing to `ipython`. It seems even `ipython` has a lot of code which has a lot of white spaces.. 

I do agree.. it seems to be a better default nonetheless.

After posting this issue, I found that even if I disable whitespaces using `let g:pymode_trim_whitespace=0`, it actually doesn't turn off white spaces.. Will try to investigate why this is so.. my `python-mode` is a bit old.. so it possible its because of that.. I had to move `python-mode` out of the load path to be able to save without removing the whitespace..

**Comment by alphaCTzo7G** [comment]
Actually my bad.. I was using `let g:pymode_trim_whitespace=0` instead of 
`let g:pymode_trim_whitespace**s**=0`.

Also looking through PEP8, it seems there is no requirement.. buts its a general guideline to not have trailing white spaces: https://www.python.org/dev/peps/pep-0008/?#id28

 `ipython` doesn't seem to respect that but `Cython` does, and I am guessing most project would probably prefer that especially in python where whitespace is a big deal..

For the projects that don't respect that.. what do you guys think of using something like 

`set list`
`set listchars=trail:·` to highlight trailing whitespaces, so users know that there are trailing whitespaces that shouldn't be removed in distributed projects?

**Comment by fmv1992** [comment]
I'm ok with both options.

**Comment by alphaCTzo7G** [comment]
I made some changes here: https://github.com/alphaCTzo7G/python-mode/commit/647980c15fded61c1f6a51ff496aa9a5c557be80

Will test it out for the next few days.. let me know what you think or if there is a more appropriate place for these changes. 

I saw that the `editor` related configs are in `pymode_options`.. so that users can turn it off if they don't want these.. so I put it there..

**Comment by fmv1992** [comment]
@diraol , I'm sorry for being away for such a long time. I think that the stale bot has served the purpose of closing very old/legacy issues from the time before we took up this project.

In other words; maybe all the issues that are opened now are relevant.

What do you think of disabling this bot?

**Comment by diraol** [comment]
Hi @fmv1992 Nice to see you around!

Yes, the bot helps a lot!

And I think we should keep it. I'm saying this because there are lots of "simple" issues that many users just created them and never got back.
I mean, the post the issue with a question/problem, I've answered, and they never gave a feedback, so I don't know if the issue was solved or if they didn't even saw the answer.

The bot is a "nice way" to close those issues without worrying or closing by myself.

Issues like this one I think we should keep it open actively (by commenting it, "upping it" or some other way of avoiding the bot to close it).

What do you think about it?

**Comment by fmv1992** [comment]
I think the reasons to keep it are stronger than the reasons to shut it down : )

Let's keep it.

**Comment by vsego** [comment]
I know that this was closed, but I think you're missing something:

> @alphaCTzo7G I do also prefer my settings with this feature enabled. I do understand the problem while contributing or working with code that is not that clear.
> I am not sure if there is any recommendation on PEP's about this.
> If there is, I am for keeping it as it is today because of that.
> If not, this seems to be a best practice, so I think it is a good idea for us to keep it like that in order to educate newcomers and other developers about keeping their code clean.
> In any case, one can always disable this option.
> What do you think?

There is a very legitimate reason to want to keep spaces at the end of the lines. Consider this:
```
This line has no space
at the end.

This line has one space 
at the end.

This line has two spaces  
at the end.
```
in some markdown editor (for example, https://jbt.github.io/markdown-editor/). The last paragraph will be in two lines, because Markdown interprets 2 spaces at the end of a line as a new line, which is useful in list items with lots of text:
```
* This is  
  a line item  
  that I want to appear split in  
  four paragraphs.
```

This is not a very common case, but the problem with your default is that it does it silently by default. By the time the user notices that their docstrings using this are messed up, it'll be waaay too late and they'll have to repair manually.

My solution before using `python-mode` was this:
```
" Kill EOL spaces interactively
function! TrimEOL()
  exec '%s/\s\+$//gce'
endfunction
autocmd BufWritePre * call TrimEOL()
```
and I believe that something similar should be the default. So, do remove them, but interactively. Those who want this stripping off or fully automatic should be able to switch it to their favourite setting.

**Comment by leycec** [comment]
**Necrobump.** Silently trimming whitespace by default is *extremely* problematic in the general case. That's bad.

As the maintainer of [Python's first O(1) runtime type checker](https://github.com/beartype/beartype), <sup>hi! check us out, yo.</sup> I routinely declare string constants intended for subsequent munging into a form suitable for dynamic evaluation as Python code. A real-world example I just stumbled over like an [ATST walker](https://starwars.fandom.com/wiki/All_Terrain_Scout_Transport) in the La Brea Tar Pits:

```
PEP_CODE_CHECK_HINT_ROOT_PREFIX = '''
        # Type-check this passed parameter or return value against this
        # PEP-compliant type hint.
        if not 
'''
```

The last line of the above string constant is suffixed by a space – and **the space is significant.** Implicitly stripping that space produces syntactically invalid Python code that fails to evaluate. If any of my glorious legion of volunteer developers open and save that file with `python-mode` and then submit a pull request, *everything blows up.* That's bad.

Admittedly, that's what tests are for, which is why we have tests. But it shouldn't have to come to that. `python-mode` should just do the Safe Thing™ out-of-the-box rather than blithely assuming that the *Mostly* Safe Thing™ that applies to the common case applies to *every* case. It doesn't.

Let's also be honest: any Vim user worth his CLI salt that globally wants automatic whitespace trimming on buffer saves *already* uses Vim configuration recipes like [this](https://vim.fandom.com/wiki/Remove_unwanted_spaces) or [this](https://vi.stackexchange.com/a/18805/16249). `python-mode` additionally doing so by default gains users who want automatic whitespace trimming nothing while costing users who do *not* want automatic whitespace trimming something. That's also bad.

In short, this is a bad default and we should all feel bad that this is still the default in 2021. :sob:
